lib/dao.js

```diff
@@ -624,6 +624,101 @@ DataAccessObject.upsert = function(data, options, cb) {
   }
   return cb.promise;
 };
+/**
+ * Performs upsert based on the instance that matches the where clause
+ . This is a PUT operation and based on
+ * non-primary key. The filter takes one key value pair and updates single matching instance.*
+ */
+DataAccessObject.upsertWithWhere = function(where, data, options, cb) {
```

Comment:
In longer term, we want to promote `patchOrCreate` instead of `upsert`/`updateOrCreate`. Can you please add an alias `patchOrCreateWithWhere = upsertWithWhere`? See how other aliases in this file are set up.